typescript/sdk/src/ism/EvmIsmModule.ts

```diff
@@ -1,59 +1,691 @@
+      multiProvider.getSigner(args.chain),
+    );
+
+    this.chainName = this.multiProvider.getChainName(this.args.chain);
+    this.domainId = this.multiProvider.getDomainId(this.chainName);
   }
 
   public async read(): Promise<IsmConfig> {
-    return await this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
+    const config =
+      typeof this.args.config === 'string'
+        ? this.args.addresses.deployedIsm
+        : await this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
+
+    return normalizeConfig(config);
   }
 
-  public async update(_config: IsmConfig): Promise<EthersV5Transaction[]> {
-    throw new Error('Method not implemented.');
+  // whoever calls update() needs to ensure that targetConfig has a valid owner
+  public async update(
+    targetConfig: IsmConfig,
+  ): Promise<AnnotatedEthersV5Transaction[]> {
+    // save current config for comparison
+    const currentConfig = await this.read();
+
+    // Update the config
+    this.args.config = targetConfig;
+
+    // moduleMatchesConfig expects any domain filtering to have been done already
+    if (
+      typeof targetConfig !== 'string' &&
+      (targetConfig.type === IsmType.ROUTING ||
+        targetConfig.type === IsmType.FALLBACK_ROUTING)
+    ) {
+      // filter for known domains
+      const { availableDomains } = this.filterRoutingIsmDomains({
+        config: targetConfig,
+      });
+      targetConfig.domains = availableDomains;
+    }
+
+    // If configs match, no updates needed
+    if (configDeepEquals(currentConfig, targetConfig)) {
+      return [];
+    }
+
+    // Else, we have to figure out what an update for this ISM entails
+
+    // If target config is a custom ISM, just update the address
+    // if config -> custom ISM, update address
+    // if custom ISM -> custom ISM, update address
+    if (typeof targetConfig === 'string') {
+      // TODO: https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3773
+      this.args.addresses.deployedIsm = targetConfig;
+      return [];
+    }
+
+    // Check if we need to deploy a new ISM
+    if (
+      // if custom ISM -> config, do a new deploy
+      typeof currentConfig === 'string' ||
+      // if config -> config, AND types are different, do a new deploy
+      currentConfig.type !== targetConfig.type ||
+      // if it is not a mutable ISM, do a new deploy
+      !MutableIsmType.includes(targetConfig.type)
+    ) {
+      const contract = await this.deploy({
+        config: targetConfig,
+      });
+
+      this.args.addresses.deployedIsm = contract.address;
+      return [];
+    }
+
+    // Explicitly narrow the type of targetConfig to RoutingIsmConfig or PausableIsmConfig
+    // This is safe because we've already checked that the type is in MutableIsmType
+    targetConfig = targetConfig as RoutingIsmConfig | PausableIsmConfig;
+
+    // At this point, only the 3 ownable/mutable ISM types remain: PAUSABLE, ROUTING, FALLBACK_ROUTING
```

Comment:
prefer switch case or add an else throw in the event that we add more types to this